### PR TITLE
Cache: new Cache format in decoder-only models

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR adds to all decoder-only models (except for XLNet) support for dynamic cache. I am working on general support for cache class in all models, but I found that encoder-decoder models are trickier and we might want to make a tuple of DynamicCache for those. That's why I am splitting the task into multiple PRs for easier review :)

All models were testes with GenerationTesterMixin tests and passed on my end. For special Bloom case, I made a tiny change that cache will hold same tensor-shapes in key/values (we need to be able to concat at shape[-2]). Another special case Git seems to work from the tests.

For encoder-decoder models I will try to add first for one model and make a PR. If the format is okay, I'll open another PR with all the generative models left

### CODE REVIEW

File: src/transformers/models/data2vec/modeling_data2vec_audio.py [hunk 1]
```diff
@@ -622,6 +622,7 @@ def _flash_attention_forward(
         """
         Calls the forward method of Flash Attention - if the input hidden states contain at least one padding token
         first unpad the input, then computes the attention scores and pad the final attention scores.
+
```
Comment (zucchini-nlp) [inline src/transformers/models/data2vec/modeling_data2vec_audio.py:625]: These kind of changes are from fix copies, and are not related at all to the PR. But let's leave it here as it's anyway related to code-consistency in the library

**Comment by gante** [comment]
> For encoder-decoder models I will try to add first for one model and make a PR. 

@zucchini-nlp regarding encoder-decoder models: let's first finish the PR for Whisper (https://github.com/huggingface/transformers/pull/31166), and then copy the design all over the library! :D Your input in that PR would be appreciated :)

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -367,29 +367,20 @@ def _crop_past_key_values(model, past_key_values, maximum_length):
...
-    elif "bloom" in model.__class__.__name__.lower() or (
-        model.config.architectures is not None and "bloom" in model.config.architectures[0].lower()
-    ):
-        for idx in range(len(past_key_values)):
-            new_past.append(
-                (
-                    past_key_values[idx][0][:, :, :maximum_length],
-                    past_key_values[idx][1][:, :maximum_length, :],
+        if isinstance(past_key_values[0], DynamicCache):
+            past_key_values[0].crop(maximum_length)
```
Comment (gante) [inline src/transformers/generation/candidate_generator.py:371]: This is related to the Whisper PR, correct? If so, this change should come in that PR, and not on this one :)

(In fact, I'm not very happy with `past_key_values[0]` -- see the Whisper PR 😛 )

File: src/transformers/generation/candidate_generator.py [hunk 3]
```diff
@@ -367,29 +367,20 @@ def _crop_past_key_values(model, past_key_values, maximum_length):
...
-            )
-        past_key_values = tuple(new_past)
-    # bloom is special
-    elif "bloom" in model.__class__.__name__.lower() or (
-        model.config.architectures is not None and "bloom" in model.config.architectures[0].lower()
-    ):
-        for idx in range(len(past_key_values)):
-            new_past.append(
-                (
-                    past_key_values[idx][0][:, :, :maximum_length],
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/candidate_generator.py:387]: If bloom changes the order of dimensions of its KV cache as a result of this PR, then this PR is a breaking change. In that case, I'd like to request to move all Bloom-related changes into a separate PR, so we can isolate the breaking changes 🎯

**Review by gante — COMMENTED** [review]
Two high-level comments to clear out before diving into details :)

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 4]
```diff
@@ -491,14 +477,23 @@ def _upad_input(self, query_layer, key_layer, value_layer, attention_mask, query
 
 
 class GPTNeoAttention(nn.Module):
-    def __init__(self, config, layer_id=0):
```
Comment (gante) [inline src/transformers/models/gpt_neo/modeling_gpt_neo.py:494]: uhmmm... I think we can leave the old `layer_id` arg as it was, as it is less BC breaking!

We wouldn't need the warning in this case.

**Review by gante — APPROVED** [review]
LGTM, thank you for working through these models 💪 

request: let's activate slow tests! (I don't trust my review to capture all nuances from all architectures 😉 )

File: src/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/gpt_neo/modeling_gpt_neo.py]: oke done

**Comment by zucchini-nlp** [comment]
All slow tests are passing on my end

File: src/transformers/models/codegen/modeling_codegen.py [hunk 5]
```diff
@@ -150,7 +158,7 @@ def _attn(
     def forward(
         self,
         hidden_states: Optional[torch.FloatTensor],
-        layer_past: Optional[Tuple[torch.Tensor]] = None,
+        layer_past: Optional[Cache] = None,
```
Comment (ArthurZucker) [inline src/transformers/models/codegen/modeling_codegen.py:161]: I think the `cache_position` are missing here as well 😅

File: src/transformers/models/codegen/modeling_codegen.py [hunk 6]
```diff
@@ -457,11 +476,12 @@ def forward(
...
-            past_length = 0
-            past_key_values = tuple([None] * len(self.h))
-        else:
-            past_length = past_key_values[0][0].size(-2)
+        past_length = 0
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_length = past_key_values.get_seq_length()
```
Comment (ArthurZucker) [inline src/transformers/models/codegen/modeling_codegen.py:484]: this is super ugly IMO. 
I would really love for us to just no add this code that ended up being copy pasted everywhere because we were lazy to update when porting new models. 

Let's use cache positions. And let's already deprecate the casting and legacy cache etc, to make sure we only do this for one revision!

**Review by ArthurZucker — COMMENTED** [review]
Short review, I want to better understand why we are shipping `get_seq_length` while we want to deprecate this API / it will make our life harder for any other cache class we want to support for these models (PS I plan to support more 😉 ) 

Let's use cache positions! 

Otherwise great work isolating the changes everywhere, might be good to even add some copied from here and there!

File: src/transformers/models/codegen/modeling_codegen.py (hunk 6 above)
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:484]: I see, I wanted to add `cache_position` as part of StaticCache support but it can be added in this PR also. Wondering about deprecation cycle for Cache class, afaik there were no deprecation warning about that before so we would still have to keep the ugly hack and add a warning message? 

I'm pro of totally getting rid of old cache, if it doesn't break BC

Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:484]: Ah I see that Llama on main had its preparation modified already assuming that the inputs is always Cache object. Oke, then it makes sense to get rid of legacy_cache in forward also

Comment (ArthurZucker) [reply src/transformers/models/codegen/modeling_codegen.py:484]: Yeah it was not done before but this PR IMO is a good time to do it! So keep this code, but add a warning saying, we are automatically converting your cache from tuple to dynamic cache class (SHould only be triggered outside generate because generate should already pass a cache class!)

Yep it makes sense but let's not be too brutal in case some people still use it! We give them one release until we totally remove it!

Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py]: Oke, added cache position and deprecation warnings to all models from this PR. I'll add the same deprecation warning to all models that already support cache class in another PR. This one is ready for review, tests are passing on my end!

**Comment by ArthurZucker** [comment]
Ping me for another review when needed! 🤗

**Comment by zucchini-nlp** [comment]
@ArthurZucker yep, ready for final review, changed the deprecation to be same as in another PR. Git model is a bit peculiar and thus will not support `cache_position`

File: src/transformers/models/codegen/modeling_codegen.py [hunk 7]
```diff
@@ -303,6 +313,7 @@ class CodeGenPreTrainedModel(PreTrainedModel):
     supports_gradient_checkpointing = True
     _no_split_modules = ["CodeGenBlock"]
     _skip_keys_device_placement = "past_key_values"
+    _supports_cache_class = True
```
Comment (gante) [inline src/transformers/models/codegen/modeling_codegen.py:316]: Don't forget 

```py
_supports_quantized_cache = True
_supports_static_cache = True
```

(if appropriate, on this and other models)

**Review by gante — COMMENTED** [review]
A nit :)

There are also two files that need to be removed (`bart.py` and `update_llava.py`)

File: src/transformers/models/codegen/modeling_codegen.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:316]: Oh, yes, forgot about `_supports_quantized_cache` flag. But I am not sure about the "static_cache" flag. Will it imply that models are fullgraph compilable, cause we didn't test it

in any case I'll make a following PR to check fullgraph compile and if it works as-is add the tests in each modeling

Comment (gante) [reply src/transformers/models/codegen/modeling_codegen.py:316]: 👍 

We should add a mixin test for fullgraph compilation when `_supports_static_cache = True`

Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:316]: Yes, I will test and add a mixin test (one lightweight and one slow maybe) after this PR is merged. 

> I think _supports_static_cache and _supports_cache_class  will mean the same thing 

Btw, GIT will be an exception which supports cache class but not static cache as it has some special attn mask preparation steps

**Comment by zucchini-nlp** [comment]
Done + added slow test commit but I don't know why it's skipped in CI

**Comment by zucchini-nlp** [comment]
Noticed that VLMs do not get dynamic cache by default, so added cache flags to VLMs + ran tests. Unfortunately quantized and static cache will not be supported until #30962 is merged

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 8]
```diff
@@ -64,6 +65,9 @@ class GPTNeoXPreTrainedModel(PreTrainedModel):
     _no_split_modules = ["GPTNeoXLayer"]
     _skip_keys_device_placement = "past_key_values"
     _supports_flash_attn_2 = True
+    _supports_cache_class = True
+    _supports_quantized_cache = True
+    _supports_static_cache = True
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:70]: cool, this is automatically tested, but manually testing would be nice to make sure we have correct outputs!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 9]
```diff
@@ -1025,22 +1137,42 @@ def forward(
         )
 
     def prepare_inputs_for_generation(
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:1139]: seems to be the same as llama: no token type ids

File: src/transformers/models/gptj/modeling_gptj.py [hunk 10]
```diff
@@ -342,18 +351,16 @@ def forward(
         query = query.permute(0, 2, 1, 3)
         # value: batch_size x num_attention_heads x seq_length x head_dim
 
+        # Note that this cast is quite ugly, but is not implemented before ROPE as the original codebase keeps the key in float32 all along the computation.
+        # Reference: https://github.com/kingoflolz/mesh-transformer-jax/blob/f8315e3003033b23f21d78361b288953064e0e76/mesh_transformer/layers.py#L128
```
Comment (ArthurZucker) [inline src/transformers/models/gptj/modeling_gptj.py:355]: not sure we need this anhymore

File: src/transformers/models/idefics/modeling_idefics.py [hunk 11]
```diff
@@ -631,18 +652,16 @@ def forward(
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            kv_seq_len += past_key_value[0].shape[-2]
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
```
Comment (ArthurZucker) [inline src/transformers/models/idefics/modeling_idefics.py:655]: should use cache positions

**Review by ArthurZucker — COMMENTED** [review]
Nice!
- let's not warn when we are training ( though I suspect a lot of people generate with a model that has model.training.... ) 
- let's add integration tests with compile 
- let's add to the compile doc, the list of supported models + performance boost 
Maybe for another PR or for @ydshieh :
- let's add these models to the list of model we run benchmark for

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 8 above)
Comment (zucchini-nlp) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:70]: You mean add slow integration tests for each model and cache type, or simply run some generations on my end to see that it makes sense?

File: src/transformers/models/falcon/modeling_falcon.py [hunk 12]
```diff
@@ -406,32 +413,34 @@ def forward(
...
-
         if self._use_sdpa and query_layer.device.type == "cuda" and attention_mask is not None:
             # For torch<=2.1.2, SDPA with memory-efficient backend is bugged with non-contiguous inputs with custom attn_mask,
             # Reference: https://github.com/pytorch/pytorch/issues/112577.
             query_layer = query_layer.contiguous()
             key_layer = key_layer.contiguous()
             value_layer = value_layer.contiguous()
 
+        if attention_mask is not None:
+            attention_mask = attention_mask[:, :, :, : key_layer.shape[-2]]
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/falcon/modeling_falcon.py:427]: Might be, but my main idea is to deprecate old cache in this PR and check if these changes enabled compile-compatibility for them. So tests for compile and other things can go for a second PR

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 9 above)
Comment (zucchini-nlp) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:1139]: seems like yes, will check and add copy statement if possible

File: src/transformers/models/gptj/modeling_gptj.py (hunk 10 above)
Comment (zucchini-nlp) [reply src/transformers/models/gptj/modeling_gptj.py:355]: yeah, seems like it was not needed as we cast it back to fp32 just after a few lines

**Comment by zucchini-nlp** [comment]
Discovered just today that GPT-Neo and a few models are just like Gemma2, i.e. have hybrid cache format. It will take me some time to figure out how to make things work and still support old format

**Comment by zucchini-nlp** [comment]
@ArthurZucker teh only model that is like Gemma2 is GPT-Neo. I started working on adding a HybridCache support on it but realized it will bloat up the code with if/elses to check on which cache type we're using now. So, I propose to review and merge this PR to deprecate cache as tuple.

For gpt-neo currently I just toggled static-cache support to `False`, and added a TODO for us. Might be quite breaking and imo should be part of compile compatibility, not deprecation PR.

**Comment by ArthurZucker** [comment]
Cool will do one last pass then!

File: src/transformers/models/codegen/modeling_codegen.py [hunk 13]
```diff
@@ -439,85 +510,61 @@ def forward(
...
+        if inputs_embeds is None:
+            inputs_embeds = self.wte(input_ids)
 
-        if past_key_values is None:
-            past_length = 0
-            past_key_values = tuple([None] * len(self.h))
-        else:
-            past_length = past_key_values[0][0].size(-2)
+        use_legacy_cache = False
+        if use_cache and not isinstance(past_key_values, Cache) and not self.training:
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/codegen/modeling_codegen.py:529]: Maybe the warning should be the only thing that check not self.training!

**Review by ArthurZucker — APPROVED** [review]
🚀 huge work and refactoring! 

Mostly a few comments on the `codegen` valid for other model:
- in generate cache positions only should be used
- not creating past is breaking, but we can break TBH
- For another PR maybe, but let's move the supports_quantized_cache to `cache_utils` with a set!

File: src/transformers/models/codegen/modeling_codegen.py [hunk 14]
```diff
@@ -591,11 +718,28 @@ def get_output_embeddings(self):
...
-        if past_key_values:
-            past_length = past_key_values[0][0].shape[2]
+        if past_key_values is not None:
+            past_length = cache_position[0] if cache_position is not None else past_key_values.get_seq_length()
+            max_cache_length = (
+                torch.tensor(past_key_values.get_max_length(), device=input_ids.device)
+                if past_key_values.get_max_length() is not None
+                else None
+            )
+            cache_length = past_length if max_cache_length is None else torch.min(max_cache_length, past_length)
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:742]: right, this code was updated in other models recently and wasn't ported to the PR

File: src/transformers/models/codegen/modeling_codegen.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:529]: Yes, can be done. Though I'm not sure how often users want to cache when training

File: src/transformers/models/idefics/modeling_idefics.py [hunk 15]
```diff
@@ -634,18 +709,17 @@ def forward(
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            kv_seq_len += past_key_value[0].shape[-2]
+            kv_seq_len += cache_position[0]
+
         if not is_cross_attention:
             cos, sin = self.rotary_emb(value_states, seq_len=max(kv_seq_len, q_len))
             query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/idefics/modeling_idefics.py:716]: Yep, RoPE standardization will go with its own PR

**Comment by zucchini-nlp** [comment]
> BTW we could also just add the class to the set _support_quantized_cache={""} in cache_utils, we don't pollute this here, and we can directly get all classes that support quantized / static etc.

Sounds good, but imo it should be another PR because this one is finalizing deprecation for a few models left. I'll make a PR for auto-mapping-cache to see how it goes

File: src/transformers/models/codegen/modeling_codegen.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/codegen/modeling_codegen.py:742]: let's port it here as well no?

**Comment by ArthurZucker** [comment]
yep another pr for sure!

File: src/transformers/models/codegen/modeling_codegen.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/codegen/modeling_codegen.py:742]: yes, on my way and running final slow tests to make sure we didn't break anything here

**Comment by zucchini-nlp** [comment]
Ran the tests and fixed in couple places.  Will merge later today/tomorrow

Btw @gante seems like compile isn't working on current `main` after making cache classes `nn.Module`. Fails to `register_buffer` for some reason. For ex: `tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_generate_compile_fullgraph` fails

**Comment by gante** [comment]
@zucchini-nlp will have a look, thank you for flagging 👍
